File: amazon/127_word_ladder.cpp

//BFS two pointers.
#include<string>
#include<vector>
#include <unordered_set>
#include<iostream>

using namespace std;

class Solution{
public: 
    int ladderLength(string beginWord, string endWord, vector<string>& wordList){
        unordered_set<string> dict(wordList.begin(), wordList.end()), head, tail,*phead, *ptail;
        if(dict.find(endWord)==dict.end()){
            return 0;
        }
        head.insert(beginWord);
        tail.insert(endWord);
        int ladder = 2;
        while(!head.empty() && !tail.empty()){
            if(head.size() < tail.size()){
                phead = &head;
                ptail = &tail;
            }else{
                phead = &tail;
                ptail = &head;
            }
            unordered_set<string> temp;
            for(auto it = phead -> begin(); it != phead -> end();it++){
                string word = *it;
                for(int i =0 ; i< word.size(); i++){
                    char t = word[i];
                    for(int j = 0; j < 26; j++){
                        word[i] = 'a' + j;
                        if(ptail->find(word) != ptail -> end()){
                            return ladder;
                        }
                        if(dict.find(word) != dict.end()){
                            temp.insert(word);
                            dict.erase(word);
                        }
                    }
                    word[i]=t;
                }
            }
            ladder++;
            phead->swap(temp);
        }
        return 0;
    }
};

int main(){
    string beginWord, endWord;
    int n;
    cout << "Enter beginWord: ";
    cin >> beginWord;
    cout << "Enter endWord: ";
    cin >> endWord; 
    cout << "Enter number of words in wordList: ";
    cin >> n;
    vector<string> wordList(n);
    cout << "Enter words in wordList:\n";
    for(int i =0; i < n ; i++){
        cin >> wordList[i];
    }
    Solution my_solution;
    int result = my_solution.ladderLength(beginWord, endWord, wordList);
    cout << "Output: "<< result << endl;
    return 0;
}

File: amazon/127_word_ladder.py
from collections import defaultdict
from collections import deque
from typing import List
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList or not endWord or not beginWord or not wordList:
            return 0
        L = len(beginWord)
        all_combo_dict = defaultdict(list)
        for word in wordList:
            for i in range(L):
                all_combo_dict[word[:i] + "*" + word[i+1:]].append(word) 
        queue = deque([(beginWord, 1)])
        visited = set()
        visited.add(beginWord)
        while queue:
            current_word, level = queue.popleft()
            for i in range(L):
                intermediate_word = current_word[:i] + "*" + current_word[i+1:]
                for word in all_combo_dict[intermediate_word]:
                    if word == endWord:
                        return level + 1
                    if word not in visited:
                        visited.add(word)
                        queue.append((word, level + 1))
        return 0

def main():
    beginWord = input("Enter beginWord: ")
    endWord = input("Enter endWord: ")
    n = int(input("Enter number of words in wordList: "))

    wordList = []
    print("Enter words in wordList: ")
    for _ in range(n):
        wordList.append(input())

    solution = Solution()
    result = solution.ladderLength(beginWord, endWord, wordList)
    print("Output:", result)

if __name__ == "__main__":
    main()